File: solutions/R/stage_3/stage3_signal_readings.R
#!/usr/bin/env Rscript

# Get command-line arguments
args <- commandArgs(trailingOnly = TRUE)

# Check if dataset path is provided
if (length(args) < 1) {
  stop("Usage: Rscript stage3_signal_readings.R <dataset_path> [plot_output_path]")
}

dataset_path <- args[1]

# Determine plot output path
if (length(args) > 1) {
  plot_path <- args[2]
} else {
  base_name <- tools::file_path_sans_ext(basename(dataset_path))
  dir_name <- dirname(dataset_path)
  plot_path <- file.path(dir_name, paste0(base_name, "_boxplot.png"))
}

# Load the dataset
data <- read.csv(dataset_path, stringsAsFactors = FALSE)

if (nrow(data) == 0) {
  stop("Dataset is empty")
}

# Group by district
districts <- sort(unique(data$district))

# Create signals list
signals <- list()
codes <- list()

for (district in districts) {
  subset_data <- data[data$district == district, ]
  signals[[district]] <- as.numeric(subset_data$signal_strength)
  
  district_code_vals <- as.integer(as.numeric(subset_data$district_code[!is.na(subset_data$district_code)]))
  if (length(district_code_vals) > 0) {
    codes[[district]] <- unique(district_code_vals)
  }
}

# One-way ANOVA + Tukey HSD post-hoc (adjusted p-values included).
anova_model <- aov(signal_strength ~ district, data = data)
tukey <- TukeyHSD(anova_model, "district")$district

pairwise_rows <- data.frame(
  pair = rownames(tukey),
  diff = tukey[, "diff"],
  p_adj = tukey[, "p adj"],
  stringsAsFactors = FALSE
)

ranked <- list()
for (district in districts) {
  district_rows <- pairwise_rows[
    grepl(sprintf("^%s-", district), pairwise_rows$pair) |
    grepl(sprintf("-%s$", district), pairwise_rows$pair),
    ,
    drop = FALSE
  ]
  if (nrow(district_rows) == 0) {
    next
  }

  sig_count <- sum(district_rows$p_adj < 0.05)
  avg_abs_diff <- mean(abs(district_rows$diff))
  med_p <- median(district_rows$p_adj)

  ranked[[district]] <- data.frame(
    sig_count = sig_count,
    avg_abs_diff = avg_abs_diff,
    med_p = med_p,
    district = district,
    stringsAsFactors = FALSE
  )
}

ranked_df <- do.call(rbind, ranked)
ranked_df <- ranked_df[order(-ranked_df$sig_count, -ranked_df$avg_abs_diff, ranked_df$med_p), ]

if (nrow(ranked_df) == 0) {
  stop("Could not identify abnormal district")
}

abnormal_district <- ranked_df$district[1]

# Get district codes
district_codes <- codes[[abnormal_district]]
district_codes <- district_codes[district_codes > 0]

if (length(district_codes) != 1) {
  stop("Abnormal district code is missing or inconsistent")
}

# Print ranking info to stderr
cat("Top post-hoc candidates (TukeyHSD-adjusted):\n", file = stderr())
for (i in seq_len(min(5, nrow(ranked_df)))) {
  cat(sprintf("  %s: significant_pairs=%d, avg_|diff|=%.3f, median_p≈%.4f\n",
              ranked_df$district[i],
              ranked_df$sig_count[i],
              ranked_df$avg_abs_diff[i],
              ranked_df$med_p[i]),
      file = stderr())
}
cat(sprintf("Identified district: %s\n", abnormal_district), file = stderr())

# Create boxplot
tryCatch({
  png(plot_path, width = max(1400, length(districts) * 60), height = 600, res = 100)
  
  data_list <- lapply(districts, function(d) signals[[d]])
  colors <- ifelse(districts == abnormal_district, "#d64545", "#5b8ff9")
  
  # Create labels with district codes
  labels <- character(length(districts))
  for (i in seq_along(districts)) {
    dist <- districts[i]
    code_vals <- codes[[dist]]
    if (!is.null(code_vals)) {
      code_vals <- sort(code_vals[code_vals > 0])
      code_label <- if (length(code_vals) > 0) as.character(code_vals[1]) else "?"
    } else {
      code_label <- "?"
    }
    labels[i] <- sprintf("%s [%s]", dist, code_label)
  }
  
  boxplot(data_list,
          names = labels,
          main = sprintf("Stage 3 Signal Pattern (identified: %s)", abnormal_district),
          ylab = "Signal Strength",
          xlab = "District",
          col = colors,
          las = 2,
          cex.axis = 0.8)
  
  # Add legend
  legend("topright",
         legend = c("Identified district", "Other districts"),
         fill = c("#d64545", "#5b8ff9"))
  
  dev.off()
}, error = function(e) {
  cat(sprintf("Warning: Could not create boxplot: %s\n", e$message), file = stderr())
})

# Output the recovery code
code_value <- district_codes[1] %% 1000000
cat(sprintf("%06d\n", code_value))

File: solutions/python/stage_3/stage3_signal_readings.py
#!/usr/bin/env python3
import csv
import math
import statistics
import sys
from collections import defaultdict
from pathlib import Path


def load_rows(path):
    with open(path, newline="", encoding="utf-8") as handle:
        return list(csv.DictReader(handle))


def group_by_district(rows):
    signals = defaultdict(list)
    codes = defaultdict(set)
    for row in rows:
        district = row.get("district")
        if not district:
            continue
        signals[district].append(float(row.get("signal_strength", 0.0)))
        if row.get("district_code"):
            codes[district].add(int(float(row["district_code"])))
    return signals, codes


def welch_stats(a, b):
    if len(a) < 2 or len(b) < 2:
        return 0.0, 1.0
    mean_a, mean_b = statistics.mean(a), statistics.mean(b)
    var_a, var_b = statistics.variance(a), statistics.variance(b)
    se = math.sqrt((var_a / len(a)) + (var_b / len(b)))
    if se == 0:
        return 0.0, 1.0
    t = abs((mean_a - mean_b) / se)
    p = math.erfc(t / math.sqrt(2.0))  # fast normal-approx two-sided p
    return t, p


def pick_abnormal_district(signals):
    districts = sorted(signals)
    ranked = []
    for district in districts:
        pairs = [welch_stats(signals[district], signals[other]) for other in districts if other != district]
        if not pairs:
            continue
        n_tests = len(pairs)
        corrected_ps = [min(1.0, p * n_tests) for _, p in pairs]  # Bonferroni correction
        sig_count = sum(1 for p in corrected_ps if p < 0.05)
        avg_abs_t = statistics.mean(t for t, _ in pairs)
        med_p = statistics.median(corrected_ps)
        ranked.append((sig_count, avg_abs_t, -med_p, district))
    if not ranked:
        return None, []
    ranked.sort(reverse=True)
    return ranked[0][3], ranked


def save_boxplot(signals, codes, abnormal_district, output_path):
    try:
        import matplotlib.pyplot as plt
        from matplotlib.patches import Patch
    except ImportError:
        return

    districts = sorted(signals)
    data = [signals[d] for d in districts]
    colors = ["#d64545" if d == abnormal_district else "#5b8ff9" for d in districts]
    pos = list(range(1, len(districts) + 1))

    fig, ax = plt.subplots(figsize=(max(14, len(districts) * 0.6), 6))
    bp = ax.boxplot(data, positions=pos, widths=0.7, patch_artist=True, showfliers=False)
    for patch, color in zip(bp["boxes"], colors):
        patch.set_facecolor(color)
        patch.set_alpha(0.7)
    ax.set_xticks(pos)
    labels = []
    for district in districts:
        code_values = sorted(code for code in codes.get(district, set()) if code > 0)
        code_label = str(code_values[0]) if code_values else "?"
        labels.append(f"{district} [{code_label}]")
    ax.set_xticklabels(labels, rotation=60, ha="right")
    ax.set_ylabel("Signal Strength")
    ax.set_xlabel("District")
    ax.set_title(f"Stage 3 Signal Pattern (identified: {abnormal_district})")
    ax.legend(
        handles=[
            Patch(facecolor="#d64545", alpha=0.7, label="Identified district"),
            Patch(facecolor="#5b8ff9", alpha=0.7, label="Other districts"),
        ],
        loc="upper right",
    )
    ax.grid(axis="y", linestyle="--", alpha=0.3)
    fig.tight_layout()
    fig.savefig(output_path, dpi=150)
    plt.close(fig)


def default_plot_path(dataset_path):
    src = Path(dataset_path)
    return str(src.with_name(f"{src.stem}_boxplot.png"))


def main():
    if len(sys.argv) < 2:
        raise SystemExit("Usage: python stage3_signal_readings.py <dataset_path> [plot_output_path]")

    dataset_path = sys.argv[1]
    plot_path = sys.argv[2] if len(sys.argv) > 2 else default_plot_path(dataset_path)
    rows = load_rows(dataset_path)
    if not rows:
        raise SystemExit("Dataset is empty")

    signals, codes = group_by_district(rows)
    abnormal_district, ranked = pick_abnormal_district(signals)
    if not abnormal_district:
        raise SystemExit("Could not identify abnormal district")

    district_codes = {c for c in codes.get(abnormal_district, set()) if c > 0}
    if len(district_codes) != 1:
        raise SystemExit("Abnormal district code is missing or inconsistent")

    print("Top post-hoc candidates (Bonferroni-corrected):", file=sys.stderr)
    for sig, avg_t, neg_med_p, district in ranked[:5]:
        print(
            f"  {district}: significant_pairs={sig}, avg_|t|={avg_t:.3f}, median_p≈{-neg_med_p:.4f}",
            file=sys.stderr,
        )
    print(f"Identified district: {abnormal_district}", file=sys.stderr)

    save_boxplot(signals, codes, abnormal_district, plot_path)
    print(f"{next(iter(district_codes)) % 1_000_000:06d}")


if __name__ == "__main__":
    main()
